lhotse/features.py

```diff
@@ -122,20 +122,56 @@ class Features:
     channel_id: int
     start: Seconds
     duration: Seconds
+
+    # The Features object must know its frame length and shift to trim the features matrix when loading.
```

Comment:
Incidentally: I hope to use the 'snip_edges: false' option for Kaldi-like feature extraction.. this makes it so that the num-frames does not depend on the frame-shift, which is a nice simplification.  We could possibly make this  requirement.

Reply:
... in any case I think we need to be clear about what exactly 'start' and 'duration' mean.
E.g. is it the 1st and last+1 sample which is *covered* by the features?  Is it the center of the 1st frame until the center of the last-plus-one frame?  Perhaps it could start at n*frame_shift.   [Note: the center of each frame would be n*frame_shift + 0.5].

Also regarding snip_edges: for consistency when operating inside the file, we could possibly do it differently for internal segments?  Maybe make snip_edges false but whenever we actually extract features for a segment, we pad with zeros or with the appropriate audio context in calling code?

Reply:
> Incidentally: I hope to use the 'snip_edges: false' option for Kaldi-like feature extraction.. this makes it so that the num-frames does not depend on the frame-shift, which is a nice simplification. We could possibly make this requirement.

I think you meant `frame-length`, right? I had to double-check with the Kaldi docs. Could you elaborate on how it makes things easier? 

> ... in any case I think we need to be clear about what exactly 'start' and 'duration' mean.

As I understand it currently, the 1st sample will start with `start`. It's less clear with `duration` - with `snip_edges = False` (default), the last sample is either at the end or after it; with `snip_edges = True`, it is either at the end or before it. You lost me at the `n*frame_shift` idea.

> Also regarding snip_edges: for consistency when operating inside the file, we could possibly do it differently for internal segments? Maybe make snip_edges false but whenever we actually extract features for a segment, we pad with zeros or with the appropriate audio context in calling code?

Yeah that's a good remark. When I add the option to extract features for specified segments I'll take care to either pass extra context samples or pad with zeros.

Reply:
frame-length, yes.
I think by `n*frame_shift`, I may have been meaning that duration == `n * frame_shift` where n == number of frames.

It would probably make sense to have `duration` always be `n * frame_shift`, but understand that this is kind of approximate and actually it may use slightly more context than that?

I'd rather make it so that downstream, these cuts behave quite simply w.r.t. the relationship of duration and num-frames, and the user doesn't have to worry about the frame-length and how it may impact the num-frames.  That was a rabbit-hole in Kaldi itself.

Reply:
OK, so you're asking that based on the `SupervisionSegment`s and the `Features`, we create the `Cuts` so that their duration is a multiple of `frame_shift`. I'm thinking we can easily do that for datasets where the recording is a stream (e.g. 15-minute phone conversation) as we have a lot of control there. For datasets where the recording is a segment already (like LibriMix and probably many others) it basically means removing a very small part of the end of the recording. Are we okay with the trade-off that we slightly modify the segment boundaries in `Cut` to have duration-num_frames consistency?